module/zfs/zio_crypt.c

```diff
@@ -0,0 +1,1987 @@
+/*
+ * CDDL HEADER START
+ *
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2017, Datto, Inc. All rights reserved.
+ */
+
+#include <sys/zio_crypt.h>
+#include <sys/dmu.h>
+#include <sys/dmu_objset.h>
+#include <sys/dnode.h>
+#include <sys/fs/zfs.h>
+#include <sys/zio.h>
+#include <sys/zil.h>
+#include <sys/sha2.h>
+
+/*
+ * This file is responsible for handling all of the details of generating
+ * encryption parameters and performing encryption and authentication.
+ *
+ * BLOCK ENCRYPTION PARAMETERS:
+ * Encryption /Authentication Algorithm Suite (crypt):
+ * The encryption algorithm, mode, and key length we are going to use. We
+ * currently support AES in either GCM or CCM modes with 128, 192, and 256 bit
+ * keys. All authentication is currently done with SHA512-HMAC.
+ *
+ * Plaintext:
+ * The unencrypted data that we want to encrypt.
+ *
+ * Initialization Vector (IV):
+ * An initialization vector for the encryption algorithms. This is
+ * used to "tweak" the encryption algorithms so that equivalent blocks of
+ * data are encrypted into different ciphertext outputs, thus obfuscating
+ * block patterns. Different modes of encryption have different requirements
+ * for the IV. AES-GCM and AES-CCM require that an IV is never reused with the
+ * same encryption key. This value is stored unencrypted and must simply be
+ * provided to the decryption function. We use a 96 bit IV (as recommended by
+ * NIST) for all block encryption. For non-dedup blocks we derive the IV
+ * randomly. The first 64 bits of the IV are stored in the second word of DVA[2]
+ * and the remaining 32 bits are stored in the upper 32 bits of blk_fill. For
+ * most object types this is safe because we only encrypt level 0 blocks which
```

Comment:
This reads a little scary: `mostly safe`, but a little broken?

How about: `This is safe because encrypted blocks can't use the upper 32 bits of blk_fill.  We only encrypt level 0 blocks, which normally have a fill count of 1.  The only exception is for DMU_OT_DNODE objects, where the fill count of level 0 blocks is the number of allocated dnodes in that block.  The on-disk format...`

Alternatively: `This is safe because for most object types we only encrypt ...`